### ISSUE TITLE
Scripting: LayerEdit.apply() doesn't trigger Automap While Drawing

### ISSUE THREAD

**Issue by eishiya** [mapeditor/tiled#4016]
Modifying a TileMap with LayerEdit.apply() doesn't trigger Automap While Drawing. Changes made with TileMap.merge() do trigger it.

Here's my code that handles click+drag painting, this is part of Tool.tilePositionChanged. Since the mouse is held down and no preview is necessary, this code paints to the layer immediately instead of creating and merging a preview.
```js
if(this.paintingDrag) {
    let line = Geometry.pointsOnLine(this.lastPosition, newPosition);
    let layer = this.map.selectedLayers[0];
    let layerEdit = layer.edit();
    layerEdit.mergeable = true; //This doesn't affect the bug, it's the same with mergeable = false too.
    let fullTile = this.getSlopeTile(0, 0, Qt.AlignBottom); //this just tells us which tile to paint
    for(let point of line) {
        layerEdit.setTile(point.x, point.y, fullTile? fullTile.tile : null, fullTile? fullTile.flags : 0);
    }
    layerEdit.apply();
}
```
Here's what the behaviour looks like, with an Automapping rule that replaces the green tile with a brown tile. In this case, one tile gets Automapped, the start of the stroke - this single tile is placed with TileMap.merge() in mousePressed, rather than LayerEdit.apply(). The tiles that remain green are the ones placed by apply() in the snippet above.
![Automapping with apply()](https://github.com/user-attachments/assets/4670921b-e2b2-4947-b56a-48afe8a2e6f8)
For comparison, here is the same green tile being drawn with Tiled's Stamp Brush tool rather than with a scripted tool. This is the behaviour I expected even with a scripted tool, all the tiles get Automapped:
![Automapping with Stamp Brush](https://github.com/user-attachments/assets/7c50c035-41d0-438a-bf6d-53c13dde761b)

**Comment by Oval17** [comment]
hey @eishiya @bjorn 
I'd like to look into this .
 would something like `layerEdit.apply(true) / layerEdit.apply(false)` be the right API like am i thinking in right direction, where true opts into emitting regionEdited? Or would a named property like `layerEdit.emitRegionEdited = true` before calling `apply() ` be preferred for readability?

**Comment by bjorn** [comment]
@Oval17 I think it should definitely be a property because the meaning of `true`/`false` is really unclear. On the other hand, `TileMap.merge` does not have the luxury of having an object to set such a state on (and hence already has a boolean `canJoin` flag). So I think we basically end up as follows:

* Add `TileLayerEdit.emitRegionEdited` (or maybe we can just call this `TileLayerEdit.notify`?). I think we can default this to `true` even if it doesn't yet do this, because that's the expected behavior.
* Add a boolean `notify` parameter to `TileMap.merge`, defaulting to `true` as well of course.
* Adjust `MapDocument::paintTileLayers` and `EditableTileLayer::applyChangesFrom` to make `emit regionEdited` conditional on this parameter. Here I think a `bool notify` parameter is fine (though it may be a little annoying to adjust all the call sites).

**Comment by eishiya** [comment]
While it's true that `true`/`false` is unclear when reading code (especially when not using code completion), inconsistency between `merge` and `apply` is also very annoying, and confusing when not using code completion. I'd personally prefer consistency in this case, as it's easier to remember one pattern than two.

+1 to defaulting both to `true` though.

Edit: Named parameter overloads are also a possibility ;D `apply({notify: false})`

**Comment by bjorn** [comment]
> I'd personally prefer consistency in this case, as it's easier to remember one pattern than two.

But it's not clear to me how we could make these functions consistent, given the existing inconsistency with the `mergeable` member vs. `canJoin` parameter. Maybe we could change both functions to take a flags parameter instead, so we can do:

```js
map.merge(stamp, EditFlags.Mergeable | EditFlags.NoNotify)
```

And similarly `edit.apply(editFlags)`. Though in that case I'm not sure how to deal with the default, since it changes based on whether it's the first time you apply an edit or whether they are repeated calls on the same edit instance... Maybe we'd just have two `apply` functions, one with that default behavior and the other using the explicit flags, though it could be a little annoying to need to set the `Mergeable` manually when one is only interested in disabling the notify.

**Comment by eishiya** [comment]
Right, I forgot about `mergeable`/`canJoin`, sorry! So currently we have `map.merge(stamp, canJoin?)` and `edit.apply()` that is modified by `edit.mergeable`.

I like the flags idea. I'd personally be fine with including a mergeable/canJoin bool or flag with all instances of `apply()` - my scripts tend to have 1-4 of each, sometimes with differing mergeability. Also, it's very rare that I need to `apply` more than once on the same TileLayerEdit, so whether `mergeable` is set on the TileLayerEdit object or in `apply`, it's about the same amount of code.

The problem seems to just be that you want a "true" flag to represent the default state, when normally flags are done the opposite way, false/0 being the default. You seem to already be fine with that with `EditFlags.NoNotify` in your example, so why not also do something like `EditFlags.NoMerge` or `EditFlags.newUndoStep`?
A `NoMerge` flag would break compatibility with existing code, but perhaps that can be resolved with a bool overload for `merge` instead of coercing the existing bool into flags? One of my scripts would benefit from such an overload anyway because it passes in a bool variable to `merge`. For `apply`, I don't think a compatibility break can be avoided, since it doesn't really make sense to have both flags and `mergeable` - `mergeable` as a default state for the flags to override doesn't really make sense.

**Comment by bjorn** [comment]
> The problem seems to just be that you want a "true" flag to represent the default state

Not necessarily, since the value of `mergeable` (and hence the default) only changes to `true` in `apply`. Generally to merge is not the default behavior, hence the flags become `Mergeable` and `NoNotify`.

I think we can rely on `true` becoming `1` for compatibility with the boolean parameter.
